### ISSUE TITLE
Self-Managed Broker fails the first time

### ISSUE THREAD

**Issue by XdoctorwhoZ** [Panduza/panduza_sandbox_flutter#8]
First when I click on Self-Managed Broker for the first time the page is greyed out. I need to go back and try again to get the two buttons available (Manual Broker and Local discovery).

**Comment by wardru** [comment]
Log message for this error:

flutter: RangeError (length): Invalid value: Not in inclusive range 0..2: 3
flutter: #0      getConnectionsButtonsList.<anonymous closure> (package:panduza_sandbox_flutter/utils_widgets/utils_widgets.dart)
flutter: #1      new ListView.separated.<anonymous closure> (package:flutter/src/widgets/scroll_view.dart:1422)
flutter: #2      SliverChildBuilderDelegate.build (package:flutter/src/widgets/scroll_delegate.dart:490)
flutter: #3      SliverMultiBoxAdaptorElement._build (package:flutter/src/widgets/sliver.dart:829)
flutter: #4      SliverMultiBoxAdaptorElement.createChild.<anonymous closure> (package:flutter/src/widgets/sliver.dart:843)
flutter: #5      BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2844)
flutter: #6      SliverMultiBoxAdaptorElement.createChild (package:flutter/src/widgets/sliver.dart:835)
flutter: #7      RenderSliverMultiBoxAdaptor._createOrObtainChild.<anonymous closure> (package:flutter/src/rendering/sliver_multi_box_adaptor.dart:349)
flutter: #8      RenderObject.invokeLayoutCallback.<anonymous closure> (package:flutter/src/rendering/object.dart:2686)
flutter: #9      PipelineOwner._enableMutationsToDirtySubtrees (package:flutter/src/rendering/object.dart:1097)
flutter: #10     RenderObject.invokeLayoutCallback (package:flutter/src/rendering/object.dart:2686)
flutter: #11     RenderSliverMultiBoxAdaptor._createOrObtainChild (package:flutter/src/rendering/sliver_multi_box_adaptor.dart:338)
flutter: #12     RenderSliverMultiBoxAdaptor.addInitialChild (package:flutter/src/rendering/sliver_multi_box_adaptor.dart:424)
flutter: #13     RenderSliverList.performLayout (package:flutter/src/rendering/sliver_list.dart:77)
flutter: #14     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #15     RenderSliverEdgeInsetsPadding.performLayout (package:flutter/src/rendering/sliver_padding.dart:139)
flutter: #16     RenderSliverPadding.performLayout (package:flutter/src/rendering/sliver_padding.dart:361)
flutter: #17     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #18     RenderViewportBase.layoutChildSequence (package:flutter/src/rendering/viewport.dart:601)
flutter: #19     RenderViewport._attemptLayout (package:flutter/src/rendering/viewport.dart:1555)
flutter: #20     RenderViewport.performLayout (package:flutter/src/rendering/viewport.dart:1464)
flutter: #21     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #22     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #23     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #24     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #25     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #26     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #27     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #28     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #29     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #30     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #31     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #32     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #33     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #34     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #35     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #36     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #37     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #38     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #39     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #40     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #41     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #42     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #43     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #44     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #45     RenderCustomPaint.performLayout (package:flutter/src/rendering/custom_paint.dart:569)
flutter: #46     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #47     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #48     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #49     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #50     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #51     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #52     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #53     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #54     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #55     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #56     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #57     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #58     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #59     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #60     RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:105)
flutter: #61     RenderObject.layout (package:flutter/src/rendering/object.dart:2575)
flutter: #62     RenderBox.layout (package:flutter/src/rendering/box.dart:2389)
flutter: #63     MultiChildLayoutDelegate.layoutChild (package:flutter/src/rendering/custom_layout.dart:173)
flutter: #64     _ScaffoldLayout.performLayout (package:flutter/src/material/scaffold.dart:1063)
flutter: #65     MultiChildLayoutDelegate._callPerformLayout (package:flutter/src/rendering/custom_layout.dart:237)
flutter: #66     RenderCustomMultiChildLayoutBox.performLayout (package:flutter/src/rendering/custom_layout.dart:404)
flutter: #67     RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:2414)
flutter: #68     PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:1051)
flutter: #69     PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:1064)
flutter: #70     RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:582)
flutter: #71     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:991)
flutter: #72     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:448)
flutter: #73     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1386)
flutter: #74     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1311)
flutter: #75     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1169)
flutter: #76     _invoke (dart:ui/hooks.dart:312)
flutter: #77     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:399)
flutter: #78     _drawFrame (dart:ui/hooks.dart:283)

**Comment by wardru** [comment]
More of a question but is it intended that when restarting the app, it kept the connection list from the previous run?

**Comment by wardru** [comment]
I tried in debug mode and there is an actual visual error, which is grayed in release mode

![Screenshot from 2024-05-03 13-18-02](https://github.com/Panduza/panduza_sandbox_flutter/assets/22248912/080b5469-81c6-4628-ab94-48f8a5cdff91)

**Comment by Darcraytore1** [comment]
I tried a lot of time to reproduce this error with windows and linux (with total reset of the application), testing on another machine but I didn't achieve to get this error on my part ... Are you sure you used the updated branch main when you did this test ?

**Comment by XdoctorwhoZ** [comment]
Maybe @wardru can list all the versions he use. Like it's flutter version, OS version. Because it seems that he have many problem that we don't and we should be able to test in its condition or indicate to users that it is not ready yet

**Comment by wardru** [comment]
```
[ 12:37 ][~/panduza/panduza_sandbox_flutter]$ flutter --version
Flutter 3.19.6 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 54e66469a9 (3 weeks ago) • 2024-04-17 13:08:03 -0700
Engine • revision c4cd48e186
Tools • Dart 3.3.4 • DevTools 2.31.1
```

```
Ubuntu 23.10 & 24.04
```

**Comment by wardru** [comment]
Tested on 2 computers with fresh cloned repo etc etc..

It seems that there is a sort of cache involved. Even when I clone and build a fresh repo, I get the connection list of the previous attempt. And on my main computer it's so buggy that I can't connect to localhost on 1883 because of the already use IP/port check. Even though the connection doesn't appear in the list and I can't delete it

**Comment by wardru** [comment]
Ok did more digging and it seems it's an issue with `shared_preferences`.

After a `rm ~/.local/share/panduza_sandbox_flutter/` the application doesn't glitch anymore.

The folder contained a `shared_preferences.json` file containing

```
{"flutter.hello":["hello","localhost","1883"],"flutter.connectionName":["hello","tryr"],"flutter.tryr":["tryr","ytyt","5","0"]}
```

It seemed to cause a bunch of issues

**Comment by wardru** [comment]
Indeed same behavior on my second PC after deleting this file.
All good now :+1: , but we should be able to explain what happend @Darcraytore1, to make sure we anitcipate these issues in the future.

After that I will merge

**Comment by Darcraytore1** [comment]
Yeah my bad, I wish I could see where it was coming. The shared preferences is used for stored the connection info on the disk and a previous version didn't delete correctly these informations, so the remaining of these informations probably cause this bug. 

But I must say I really didn't remember in which version I added "flutter.tryr" in the shared preferences even with my tests (I have checked not automatic operation added any topic in this file). I have also checked and in the current version there not any occurences of "flutter.tryr" or any adding in the shared preferences except the ones I use, so there shouldn't be any more problems of this type in the future, but in doubt I'll have to write down somewhere in the documentation the use of this file and a potential bug if the content isn't correct.
